contracts/SpeedMarkets/SpeedMarketsAMM.sol

```diff
@@ -0,0 +1,518 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+// external
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/math/MathUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/math/SafeMathUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+
+import "@openzeppelin/contracts-4.4.1/proxy/Clones.sol";
+
+import "@pythnetwork/pyth-sdk-solidity/IPyth.sol";
+import "@pythnetwork/pyth-sdk-solidity/PythStructs.sol";
+
+// internal
+import "../utils/proxy/solidity-0.8.0/ProxyReentrancyGuard.sol";
+import "../utils/proxy/solidity-0.8.0/ProxyOwned.sol";
+import "../utils/proxy/solidity-0.8.0/ProxyPausable.sol";
+import "../utils/libraries/AddressSetLib.sol";
+
+import "../interfaces/IStakingThales.sol";
+import "../interfaces/IMultiCollateralOnOffRamp.sol";
+
+import "./SpeedMarket.sol";
+
+/// @title An AMM for Thales speed markets
+contract SpeedMarketsAMM is Initializable, ProxyOwned, ProxyPausable, ProxyReentrancyGuard {
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+    using AddressSetLib for AddressSetLib.AddressSet;
+
+    AddressSetLib.AddressSet internal _activeMarkets;
+    AddressSetLib.AddressSet internal _maturedMarkets;
+
+    uint private constant ONE = 1e18;
+
+    IERC20Upgradeable public sUSD;
+
+    address public speedMarketMastercopy;
+
+    uint public safeBoxImpact;
+    uint public lpFee;
+
+    address public safeBox;
+
+    mapping(bytes32 => bool) public supportedAsset;
+
+    uint public minimalTimeToMaturity;
+    uint public maximalTimeToMaturity;
+
+    uint public minBuyinAmount;
+    uint public maxBuyinAmount;
+
+    mapping(bytes32 => uint) public maxRiskPerAsset;
+    mapping(bytes32 => uint) public currentRiskPerAsset;
+
+    mapping(bytes32 => bytes32) public assetToPythId;
+
+    //eth 0xff61491a931112ddf1bd8147cd1b641375f79f5825126d665480874634fd0ace
+    IPyth public pyth;
+
+    uint64 public maximumPriceDelay;
+
+    /// @return The address of the Staking contract
+    IStakingThales public stakingThales;
+
+    mapping(address => AddressSetLib.AddressSet) internal _activeMarketsPerUser;
+    mapping(address => AddressSetLib.AddressSet) internal _maturedMarketsPerUser;
+
+    struct MarketData {
```

Comment:
better to place `struct MarketData` on top before/after the constant. It's not a storage variable, but it's in the middle

Reply:
afaik structs can not be moved due to OpenZeppelin proxy rules, which is why it was added at the end during development and prod testing